Source/Core/DolphinQt2/Translation.cpp

```diff
@@ -0,0 +1,298 @@
+  }
+  MoIterator& operator-=(difference_type n)
+  {
+    advance(-n);
+    return *this;
+  }
+  friend MoIterator operator+(difference_type n, const MoIterator& it) { return it + n; }
+  friend MoIterator operator+(const MoIterator& it, difference_type n)
+  {
+    MoIterator tmp(it);
+    tmp += n;
+    return tmp;
+  }
+  difference_type operator-(const MoIterator& other) const { return other.distance_to(*this); }
+  friend MoIterator operator-(difference_type n, const MoIterator& it) { return it - n; }
+  friend MoIterator operator-(const MoIterator& it, difference_type n)
+  {
+    MoIterator tmp(it);
+    tmp -= n;
+    return tmp;
+  }
+
+private:
+  const char* m_data;
+  u32 m_table_offset;
+  u32 m_index;
+};
+
+class MoFile
+{
+public:
+  MoFile() = default;
+  explicit MoFile(const std::string& filename)
+  {
+    File::IOFile file(filename, "rb");
+    m_data.resize(file.GetSize());
+    file.ReadBytes(m_data.data(), m_data.size());
+
+    if (!file)
+    {
+      ERROR_LOG(COMMON, "Error reading MO file '%s'", filename.c_str());
+      m_data = {};
+      return;
+    }
+
+    u32 magic = ReadU32(&m_data[0]);
+    if (magic != MO_MAGIC_NUMBER)
+    {
+      ERROR_LOG(COMMON, "MO file '%s' has bad magic number %x\n", filename.c_str(), magic);
+      m_data = {};
+      return;
+    }
+
+    u16 version_major = ReadU16(&m_data[4]);
+    if (version_major > 1)
+    {
+      ERROR_LOG(COMMON, "MO file '%s' has unsupported version number %i", filename.c_str(),
+                version_major);
+      m_data = {};
+      return;
+    }
+
+    m_number_of_strings = ReadU32(&m_data[8]);
+    m_offset_original_table = ReadU32(&m_data[12]);
+    m_offset_translation_table = ReadU32(&m_data[16]);
+  }
+
+  u32 GetNumberOfStrings() const { return m_number_of_strings; }
+  const char* Translate(const char* original_string) const
+  {
+    const MoIterator begin(m_data.data(), m_offset_original_table);
+    const MoIterator end(m_data.data(), m_offset_original_table, m_number_of_strings);
+    auto iter = std::lower_bound(begin, end, original_string,
+                                 [](const char* a, const char* b) { return strcmp(a, b) < 0; });
+
+    if (strcmp(*iter, original_string) != 0)
+      return original_string;
+
+    u32 offset = ReadU32(&m_data[m_offset_translation_table + std::distance(begin, iter) * 8 + 4]);
+    return &m_data[offset];
+  }
+
+private:
+  std::vector<char> m_data;
+  u32 m_number_of_strings = 0;
+  u32 m_offset_original_table = 0;
+  u32 m_offset_translation_table = 0;
+};
+
+class MoTranslator : public QTranslator
+{
+public:
+  using QTranslator::QTranslator;
+
+  bool isEmpty() const override { return m_mo_file.GetNumberOfStrings() == 0; }
+  bool load(const std::string& filename)
+  {
+    m_mo_file = MoFile(filename);
+    return !isEmpty();
+  }
+
+  QString translate(const char* context, const char* source_text,
+                    const char* disambiguation = Q_NULLPTR, int n = -1) const override
```

Comment:
this can just be `nullptr`, since we don't need to worry about being C++98 compatible.

Reply:
Done.